src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt

```diff
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
 	}
 
 	private fun setProjectVersion(project: Project, versionInfo: VersionInfo) {
-		project.version = versionInfo.toString()
+		project.version = configuration.customVersionToString.apply(versionInfo)
 	}
 
 	private fun setJarManifestVersion(project: Project, versionInfo: VersionInfo) {
 		val closure = object {
 			@Suppress("unused")
 			fun apply(jar: Jar) {
 				val attributes = jar.manifest?.attributes ?: throw Exception("The Jar task has no manifest.")
-				attributes.put("Implementation-Version", versionInfo.toString())
+				attributes.put("Implementation-Version", configuration.customVersionToString.apply(versionInfo))
 				attributes.put("Implementation-Sha", versionInfo.sha)
 				attributes.put("Specification-Version", "${versionInfo.major}.${versionInfo.minor}")
 			}
 		}
 		(project.tasks ?: throw Exception("The project has no tasks.")).withType(Jar::class.java, MethodClosure(closure, "apply"))
 	}
 
-	private fun exposeVersionInfoToUser(project: Project, versionInfo: VersionInfo) {
-		project.extensions.create("ZoltuGitVersioning", Extension::class.java, versionInfo)
-	}
+	private fun exposeConfigurationObject(project: Project, configurationChangeCallback: () -> Unit) = project.extensions.create("ZoltuGitVersioning", Configuration::class.java, configurationChangeCallback)
 
-	open class Extension(open val versionInfo: VersionInfo)
+	open class Configuration(private val configurationChangeCallback: () -> Unit) {
+		companion object {
+			@Suppress("unused")
+			const val NAME = "versioning"
+			private val defaultVersionToString = Function<VersionInfo, String> { versionInfo: VersionInfo ->
+				val suffix = if (versionInfo.tags != null && versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.tags}.${versionInfo.commitCount}"
+				} else if (versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.commitCount}"
+				} else if (versionInfo.tags != null) {
+					// simple versioning
+					"${versionInfo.commitCount}-${versionInfo.tags}"
+				} else {
+					// simple versioning
+					"${versionInfo.commitCount}"
+				}
+				"${versionInfo.major}.${versionInfo.minor}.$suffix".toString()
+			}
+			private val defaultCustomDescribeProcessor = Function<String, VersionInfo?> { null }
+		}
+		/**
+		 * Contains version information for consumption.
+		 */
+		lateinit var versionInfo: VersionInfo
+
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```

Comment:
> Out of curiosity, what is the use case that you are looking to provide for? Presumably you created this PR to allow you to solve a particular problem, maybe if I understood that it would help understand the best solution for you?

I originally made this PR because I wanted to use a custom implementation of `VersionInfo.toString()` and I didn't realize until later that I could have just done something like:

```groovy
version "${ZoltuGitVersioning.versionInfo.major}.${ZoltuGitVersioning.versionInfo.minor}" +
        ".${ZoltuGitVersioning.versionInfo.patch}.${ZoltuGitVersioning.versionInfo.commitCount}"
```

Oh well :grimacing:

> What is the "usual place" you are referring to? You mean in the buildscript block? Is it possible for a plugin to expose a class to the rest of Gradle such that IntelliJ picks it up? If so, that seems like the best option.

```groovy
repositories {
    maven {
        url 'https://plugins.gradle.org/m2/'
    }
}

dependencies {
    compile group: 'gradle.plugin.com.zoltu.gradle.plugin',
            name: 'git-versioning',
            version: '2.0.34'
}
```

But this does mean that we're building an unnecessary extra dependency, which some users might not like.